pkg/querier/frontend2/frontend_scheduler_worker.go

```diff
@@ -0,0 +1,303 @@
+	// Because we only use single long-running method, it doesn't make sense to inject user ID, send over tracing or add metrics.
+	opts, err := f.cfg.GRPCClientConfig.DialOption(nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	conn, err := grpc.DialContext(ctx, address, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return conn, nil
+}
+
+// Worker managing single gRPC connection to Scheduler. Each worker starts multiple goroutines for forwarding
+// requests and cancellations to scheduler.
+type frontendSchedulerWorker struct {
+	log log.Logger
+
+	conn          *grpc.ClientConn
+	concurrency   int
+	schedulerAddr string
+	frontendAddr  string
+
+	// Context and cancellation used by individual goroutines.
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	// Shared between all frontend workers.
+	requestCh <-chan *frontendRequest
+
+	// Cancellation requests for this scheduler are received via this channel. It is passed to frontend after
+	// query has been enqueued to scheduler.
+	cancelCh chan uint64
+}
+
+func newFrontendSchedulerWorker(conn *grpc.ClientConn, schedulerAddr string, frontendAddr string, requestCh <-chan *frontendRequest, concurrency int, log log.Logger) *frontendSchedulerWorker {
+	w := &frontendSchedulerWorker{
+		log:           log,
+		conn:          conn,
+		concurrency:   concurrency,
+		schedulerAddr: schedulerAddr,
+		frontendAddr:  frontendAddr,
+		requestCh:     requestCh,
+		cancelCh:      make(chan uint64),
+	}
+	w.ctx, w.cancel = context.WithCancel(context.Background())
+
+	return w
+}
+
+func (w *frontendSchedulerWorker) start() {
+	client := NewSchedulerForFrontendClient(w.conn)
+
+	for i := 0; i < w.concurrency; i++ {
+		w.wg.Add(1)
+		go func() {
+			defer w.wg.Done()
+			w.runOne(w.ctx, client)
+		}()
+	}
+}
+
+func (w *frontendSchedulerWorker) stop() {
+	w.cancel()
+	w.wg.Wait()
+	if err := w.conn.Close(); err != nil {
+		level.Error(w.log).Log("msg", "error while closing connection to scheduler", "err", err)
+	}
+}
+
+func (w *frontendSchedulerWorker) runOne(ctx context.Context, client SchedulerForFrontendClient) {
+	backoffConfig := util.BackoffConfig{
+		MinBackoff: 50 * time.Millisecond,
+		MaxBackoff: 1 * time.Second,
+	}
+
+	backoff := util.NewBackoff(ctx, backoffConfig)
+	for backoff.Ongoing() {
+		loop, loopErr := client.FrontendLoop(ctx)
+		if loopErr != nil {
+			level.Error(w.log).Log("msg", "error contacting scheduler", "err", loopErr, "addr", w.schedulerAddr)
+			backoff.Wait()
+			continue
+		}
+
+		loopErr = w.schedulerLoop(ctx, loop)
+		if closeErr := loop.CloseSend(); closeErr != nil {
+			level.Debug(w.log).Log("msg", "failed to close frontend loop", "err", loopErr, "addr", w.schedulerAddr)
+		}
+
+		if loopErr != nil {
+			level.Error(w.log).Log("msg", "error sending requests to scheduler", "err", loopErr, "addr", w.schedulerAddr)
+			backoff.Wait()
+			continue
```

Comment:
Yes. It can restart with the same address, and worker should then reestablish connection and start new `FrontendLoop`.

If scheduler restarts with a different address, DNS watcher will call `AddressRemoved` which will eventually stop this method, and `AddressAdded` will then start a new loop.